### PR TITLE
Node info

### PR DESCRIPTION
**Pull request by bjack205**
This adds the ability to publish a "heartbeat" from each node, containing basic information about the node. The next step (a future PR) is to write a subscriber node that takes a list of the port addresses for these messages and prints a summary of all nodes on the network.

See the end of `node_tests.jl` for an example of using this new functionality.

### CODE REVIEW

File: test/node_tests.jl [hunk 1]
```diff
@@ -149,4 +158,25 @@ dz = node.test_msg.z - subnode.test_msg.z
 @test abs(dx) <= 1
 @test abs(dy) <= 2
 @test abs(dz) <= 3
-@show node.test_msg.x
\ No newline at end of file
+@show node.test_msg.x
+end
+
+## Test heartbeat
+@testset "Heartbeat" begin
```
Comment (bjack205) [inline test/node_tests.jl:165]: Here is the example of how to use the heartbeat. You have to enable it via the node options, and then it's as simple as creating a subscriber to the correct port.

File: src/subscribers/abstract_subscriber.jl [hunk 2]
```diff
@@ -36,12 +36,13 @@ function got_new!(sub::Subscriber)
...
-    io = seek(convert(IOStream, bin_data), 0)
-    ProtoBuf.readproto(io, buf)
+function decode!(buf::ProtoBuf.ProtoType, bin_data::IOBuffer)
+    # io = IOBuffer(bin_data)
+    # io = seek(convert(IOStream, bin_data), 0)
+    ProtoBuf.readproto(bin_data, buf)
 end
 
-function decode!(buf::AbstractVector{UInt8}, bin_data)
+function decode!(buf::AbstractVector{UInt8}, bin_data::IOBuffer)
```
Comment (aabouman) [inline src/subscribers/abstract_subscriber.jl:45]: See review of  [Nonblocking receive PR](https://github.com/RoboticExplorationLab/Mercury.jl/pull/18#pullrequestreview-772766796).

File: src/utils.jl [hunk 3]
```diff
@@ -54,7 +54,14 @@ function gensubscribername()
     return name
 end
 
-
+DEFAULT_PORT = 6000;
+reset_default_port() = global DEFAULT_PORT = 6000;
```
Comment (aabouman) [inline src/utils.jl:58]: So default port numbering is cool but then its a bit clunky for the user to go to another file which contains a subscriber and get the port of the associated publisher.

File: src/node.jl [hunk 4]
```diff
@@ -1,6 +1,12 @@
 
 Base.@kwdef mutable struct NodeOptions
     rate::Float64 = 100
+    default_addr::Sockets.IPv4 = ip"127.0.0.1"
+    heartbeat_enable::Bool = false
+    heartbeat_addr::Sockets.IPv4 = default_addr
+    heartbeat_port::Int = getdefaultport()
+    heartbeat_rate::Float64 = 1.0  # Hz
+    heartbeat_print_rate_enable::Bool = false
```
Comment (aabouman) [inline src/node.jl:9]: It feels a bit weird for these heartbeat options to be inside of `NodeOptions` rather than simply a part of `NodeHeartBeat`. Why not make `NodeHeartBeat` inherit from `Publisher` and contain all of these options directly in `NodeHeartBeat`.

**Review by aabouman — COMMENTED** [review]
The heartbeat seems a bit weirdly integrated into the node and makes some of the code a bit hard to parse. The mechanics of the heartbeat in general shouldn't be super exposed to the user (in particular the user shouldn't specify the port for the pub sub). I think something more akin to a BrainNode <: Node and the user has the option of launching at a specific port which many of the HeartBeat <: Publishers inside of the node can then message. Especially cause ZMQ allows a many pub -> one sub structure.

File: src/node.jl (hunk 4 above)
Comment (bjack205) [reply src/node.jl:9]: Having options spread around everywhere in different places makes it harder to find. I my mind the heartbeat is directly tired to the node, so enabling/disabling it should go with the node options. Having one place to go to set all your options makes life much simpler.

File: src/utils.jl (hunk 3 above)
Comment (bjack205) [reply src/utils.jl:58]: I agree, I haven't figured out a good way to automatically set up the ports. But it's not too bad to extract out the port and address from the options of a node if you want to create a subscriber to the heartbeat of that node. It's be pretty easy to make a function to do that for you.

**Comment by bjack205** [comment]
I haven't done much digging into the other ZMQ messaging frameworks, but don't you need to specify the address and port regardless of which method you choose? Each node doesn't have a specific address and port (since it's publishing and subscribing to potentially many), so you need a way to specify the port for the heartbeat, right? Since the heartbeat is inherently tied to a node, I think it's natural to have it live inside the node. 

As-is, the heart-beat is pretty transparent to the user. The only thing the user has to worry about are the options in `NodeOptions`. Again, if there's a way to reduce it down to a simple enable/disable flag, that'd be great but I didn't see the path on how to do that.
